application.py

```diff
@@ -1308,6 +1361,106 @@ def get(self):
+            }
+        }
+    })
+    def get(self):
+        global e
+        e = ExpressionMatrix(application.config["DATA_DIR"], True)
+        try:
+            qs = parse_querystring(request.args)
+        except QueryStringError as error:
+            return make_payload({}, str(error), 400)
+        output_cellset = "outcellset_{}".format(request.query_string)
+        if not len(qs):
+            output_cellset = "AllCells"
+        try:
+            # create cellset
+            cellset_from_query(qs, output_cellset)
+            expression = e.getDenseExpressionMatrix("AllGenes", output_cellset)
+            expression = expression.transpose()
+            genes = all_genes()
+            cellids = e.getCellSet(output_cellset)
+            # transform to file
+            si = io.StringIO()
+            writer = csv.writer(si)
+            writer.writerow([""] + [get_cell_name(cid) for cid in cellids])
+            for idx, row in enumerate(expression):
+                writer.writerow([genes[idx]] + row.tolist())
+            return make_csv(si.getvalue(), "cxg_expression_matrix.csv")
+        except Exception as error:
+            print("ERROR creating csv", error)
+            return make_payload([], "Error: creating csv download", 400)
+        finally:
+            if output_cellset != "AllCells":
+                e.removeCellSet(output_cellset)
+
+
+class MetadataEgressAPI(Resource):
+    @swagger.doc({
+        'summary': 'Download metadata for filter based on metadata filter',
+        'tags': ['download'],
+        'description': "Cells takes query parameters defined in the schema retrieved from the /initialize enpoint. "
+                       "<br>For categorical metadata keys filter based on `key=value` <br>"
+                       " For continuous metadata keys filter by `key=min,max`<br> Either value "
+                       "can be replaced by a \*. To have only a minimum value `key=min,\*`  To have only a maximum "
+                       "value `key=\*,max` <br>Graph data (if retrieved) is normalized"
+                       " To only retrieve cells that don't have a value for the key filter by `key`",
+        'responses': {
+            '200': {
+                'description': 'csv of metadata for cells',
+            },
+
+            '400': {
+                'description': 'bad query params or bad download',
+            }
+        }
+    })
+    def get(self):
+        global e
+        e = ExpressionMatrix(application.config["DATA_DIR"], True)
+        try:
+            qs = parse_querystring(request.args)
+        except QueryStringError as error:
+            return make_payload({}, str(error), 400)
+        output_cellset = "outcellset_{}".format(request.query_string)
+        if not len(qs):
+            output_cellset = "AllCells"
+        try:
+            # create cellset
+            cell_ids = cells_from_query(qs, output_cellset)
+            metadata = parse_metadata(cell_ids)
```

Comment:
This is another spot where you may consider conversion to CSV incrementally (row-by-row) if the total number of cells is going to be large.   It would use far less memory, which will be important down the road if the back-end needs to support very large data sets and/or concurrent users.    It will require a bit of refactoring, which may be simplified by using an iterator (if you want to hide some of the initialization and state, and by nicely Pythonic).